### DISCUSSION TITLE
Welcome to modbusTCP-DotNET Discussions!

### DISCUSSION THREAD

**Discussion by stephan1827** [stephan1827/modbusTCP-DotNET#7] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Abdelazizkrichen** [comment]
Hello,

I am new to modbus, am a c# programmer and i am on a mission to make a RX scanner communicate with an application with only RJ 45 cable (direct connexion no switch) i connected the pc ( fixed IP Address with a dedicated network card a " second one") with RX card directly and i've launched your projet (first times it crashed before fixing the IP but when i did it did not crash) though it did not bring any numbers or values except 0's.

Could you please help me what am i doing wrong or should it be essentially a switch between terminals ?

Thank you.

**Reply by stephan1827** [reply]
First of all the application should not crash when you enter a false IP address. Can you tell me what exactly you entered when it crashed? Did you confirm that the modbus addresses are readable with another program? What are you trying to read, analog or digital data?

**Reply by Abdelazizkrichen** [reply]
At firs i entered a false IP, now it is readable and i pinged it it works well.
what am i trying to read ... is the question probably digital data since it is an X-Ray generater card that used to have like digital screen like any hostpital scanner ... now i want it's values (kvp, mAs ) to fill some parameters with so that it could work...

**Reply by stephan1827** [reply]
I am sorry but there are thousands of devices out there, and they all use their own addressing. There is no way for me to tell what address you find your data. This should be in the manual of the machine. You try some free modbus tools to find the addresses you are looking for before writing your own code. Here is free client you can try

http://www.simplymodbus.ca/TCPclient.htm

**Reply by Abdelazizkrichen** [reply]
Thank you it could really help me :)

**Comment by simoni88** [comment]
Hi.

Thanks to this source code, I made a modbusTCP program.  However, there is one problem..   return "144"  in function when I write multiple register...   The normal result is "16".  Why are "144" returned to function..? 
I haven't solved this problem for a month (T_T)

**Reply by simoni88** [reply]
My parameter setting value is  unit = 1 , start adrress= 2511, size = 20  and write it in words.

**Reply by timbnr** [reply]
Response of FC144 indicates a modbus exception response for FC16 requests.  The exception code will give you more details on why the device is creating the exception.  For example, address 2511 may be an invalid address on the specific server device you're talking with.

More info can be found at this link https://www.simplymodbus.ca/exceptions.htm

**Reply by stephan1827** [reply]
> My parameter setting value is unit = 1 , start adrress= 2511, size = 20 and write it in words.

Could it be that the address is in HEX? Please try to use the sample application and then read the same address. The post a screenshot of the result.

**Reply by simoni88** [reply]
Im already use your sample program.. and it worked well..  I'm not sure what's the difference.  
Im gonna try the address in hex what you said

**Comment by DrMoller** [comment]
Hi Stephan
 thanks for you code ... it has been really helpful to get our Modbus connection up running fast
Just a question. Could it be that the code in FrmStart.cs line 693
if ((x <= data.GetUpperBound(0)) && (ctrl.Text != ""))
should be
if ((x <= word.GetUpperBound(0)) && (ctrl.Text != ""))
Regards Peter

**Reply by stephan1827** [reply]
Since all arrays are the same size it does not make a difference, but from clean code perceptive you are correct. I updated the code with correct variable.

Thx

Stephan

**Comment by Saro-DevDozer** [comment]
Hi Stephen, First of all, Thank you for this work.
 I'm connecting with TCP/IP with three slaves connecting in the same IP, foreach second, I'm trying to read values in holding register from all slaves. but from only two slaves I'm getting receive_data response, the third slave is not responding back for ReadSingleRegister request, tried with multiple TCP/IP devices as well, 
 When I disable reading of any one of these 3 slaves, other 2 responds back fine.
 I'm able to read using other modscan or modbus poll applications.

**Reply by stephan1827** [reply]
Hi Saro,
can you send some code snippet how you connect the slaves? You say that you use the same IP, I guess you mean IP range? Are you creating 3 instances of the master or do you connect one by one to each slave?

**Reply by Saro-DevDozer** [reply]
Three slaves are connected in same IP. I'm creating one master with IP and port and then reading holding registers one by one using different slave id per second

**Reply by stephan1827** [reply]
Do you have OnException event? Is that triggered? Maybe the slave is not responding fast enough. What happens when you increase time from 1 second to 10 seconds?

**Reply by Saro-DevDozer** [reply]
OnException event is not triggering. Receive_Data triggered for that 3rd slave after a very long time, but the other two slave responding normal, I have tried with different devices, using other modscan apps can see all of them responding per second. Haven't tried with 10 seconds, because i need to display real-time data.

**Reply by stephan1827** [reply]
So when talk to device 1 and 2 it works fine, 1 and 3 works fine, 2 and 3 works fine as long as its just two devices at a time. But when you talk to device 1,2 and 3 the 3rd device does not respond anymore? Can you measure the time it takes for the slave to respond to one request?

Another thing you could try is to create 3 master instances that each read only from one unit. This would create 3 separate connections. It would be interesting to see what happens in this case.

**Reply by Saro-DevDozer** [reply]
Right. Any of the two devices I'm able to read holding registers rapidly, when the third device tried to read register, no response. 

I have tried with instantiating master for each slave, works fine. Thanks much.

Have another query,  when i try to write holding register with float values with decimals, only integer gets updated in the register of the slave. I can see bytes converted for decimal values. Writing value with decimals is supported?

**Reply by stephan1827** [reply]
I am still not sure why it would not work with one instance but using 3 instances might be a good workaround. In general, modbus only knows bools (coils) and 2 byte integers (registers). This is due to the history of modbus. What many devices do is to use 2 registers (4 bytes) to implement large integers or decimal values. In other words, modbus is only the transport layer. What the master and slave make out of this data is completely up to them. 

The tricky part with decimal values in 2 registers is always the byte alignment. This can be little or big endian depending on the device specs. The best way to figure this out it is to read a decimal value from the device and then put the 4 bytes from the 2 registers into a byte array. Generate the same decimal value in C# and also put it into a byte array. When you know look at the two arrays, you should see what byte has to go where in C# to generate the correct decimal value on the device.

**Reply by Saro-DevDozer** [reply]
Any sample code if you can provide for reading data from this method (MBmaster_OnResponseData(ushort ID, byte unit, byte function, byte[] data)), that would be helpful.

Currently I'm following code to read from register, which does not convert decimals properly

        private static void MBmaster_OnResponseData(ushort ID, byte unit, byte function, byte[] data)
        {
            if (ID == 7)
            {
                byte[] value = new byte[4] { 0, 0, data[0], data[1] };
                var word = ModbusUtility.NetworkBytesToHostUInt16(value);
                float final = ModbusUtility.GetSingle(word[1], word[0]);

                Console.WriteLine($"Read Updated Value: {final}");
            }
       }

**Reply by stephan1827** [reply]
I am sorry but I dont have code for that. It also depends on how the decimal is defined on the slave. Here are some code snippets that might help you.

```       // ------------------------------------------------------------------------
        // Convert byte to word
        internal int Byte2Word(byte byte1, byte byte2)
        {
            byte[] bytes = { byte2, byte1 };

            return BitConverter.ToInt16(bytes, 0);
        }

        // ------------------------------------------------------------------------
        // Convert byte to long
        internal long Byte2Long(byte byte1, byte byte2, byte byte3, byte byte4)
        {
            byte[] bytes = { byte4, byte3, byte2, byte1 };

            return BitConverter.ToInt32(bytes, 0);
        }

        // ------------------------------------------------------------------------
        // Convert byte array to word array
        internal int[] ByteArray2WordArray(byte[] values)
        {
            int[] result = new int[values.Length / 2];
            for (int x = 0; x < values.GetUpperBound(0); x = x + 2)
            {
                result[x / 2] = Byte2Word(values[x], values[x + 1]);
            }
            return result;
        }

        // ------------------------------------------------------------------------
        // Convert word array to byte array
        internal byte[] WordArray2WordByte(int[] values)
        {
            byte[] result = new byte[values.Length * 2];
            for (int x = 0; x < values.Length; x++)
            {
                byte[] _val = BitConverter.GetBytes((short)IPAddress.HostToNetworkOrder((short)values[x]));
                result[2 * x] = _val[0];
                result[2 * x + 1] = _val[1];
            }
            return result;
        }
```

**Comment by DennyInTaiwan** [comment]
Can this library be used in RTU over TCP?  thanks.

**Reply by stephan1827** [reply]
I am sorry but this is for modbusTCP only. If you are looking for a modbusRTU to TCP wrapper this will not work.

**Comment by DennyInTaiwan** [comment]
Thanks.

Stephan Stricker ***@***.***>於 2023年8月5日 週六，下午1:40寫道：

> I am sorry but this is for modbusTCP only. If you are looking for a
> modbusRTU to TCP wrapper this will not work.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stephan1827/modbusTCP-DotNET/discussions/7#discussioncomment-6643985>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBW3N55VD5FSMKSZ73QQBVDXTXMGTANCNFSM4US7NLFQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>
-- 
Denny
